packages/browser/__tests__/Session.spec.ts

```diff
@@ -0,0 +1,205 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+// The following is required by tsyringe
+import "reflect-metadata";
+import { it, describe } from "@jest/globals";
+import { mockClientAuthentication } from "../src/__mocks__/ClientAuthentication";
+import { Session } from "../src/Session";
+import { mockStorage } from "../../core/src/storage/__mocks__/StorageUtility";
+
+describe("Session", () => {
+  describe("constructor", () => {
+    it("accepts an empty config", async () => {
+      const mySession = new Session({});
+      expect(mySession.info.isLoggedIn).toEqual(false);
+      expect(mySession.info.sessionId).toBeDefined();
+    });
+
+    it("accepts no config", async () => {
+      const mySession = new Session();
+      expect(mySession.info.isLoggedIn).toEqual(false);
+      expect(mySession.info.sessionId).toBeDefined();
+    });
+
+    it("does not generate a session id if one is provided", () => {
+      const mySession = new Session({}, "mySession");
+      expect(mySession.info.sessionId).toEqual("mySession");
+    });
+
+    it("accepts input storage", async () => {
+      const insecureStorage = mockStorage({});
+      const secureStorage = mockStorage({});
+      const mySession = new Session({
+        insecureStorage,
+        secureStorage,
+      });
+      const clearSecureStorage = jest.spyOn(secureStorage, "delete");
+      const clearInsecureStorage = jest.spyOn(insecureStorage, "delete");
+      await mySession.logout();
+      expect(clearSecureStorage).toHaveBeenCalled();
+      expect(clearInsecureStorage).toHaveBeenCalled();
+    });
+
+    it("accepts session info", () => {
+      const mySession = new Session({
+        sessionInfo: {
+          sessionId: "mySession",
+          isLoggedIn: false,
+          webId: "https://some.webid",
+        },
+      });
+      expect(mySession.info.isLoggedIn).toEqual(false);
+      expect(mySession.info.sessionId).toEqual("mySession");
+      expect(mySession.info.webId).toEqual("https://some.webid");
+    });
+  });
+
+  describe("login", () => {
+    it("wraps up ClientAuthentication login", async () => {
+      const clientAuthentication = mockClientAuthentication();
+      const clientAuthnLogin = jest.spyOn(clientAuthentication, "login");
+      const mySession = new Session({ clientAuthentication });
+      await mySession.login({});
+      expect(clientAuthnLogin).toHaveBeenCalled();
```

Comment:
Again, I'd expect the mocked session `login()` method to set the `isLoggedIn` flag to true, so shouldn't this test be asserting on that flag, rather than the fact that some method on an internal instance (i.e. `clientAuthentication`'s `login` method was invoked)?

Reply:
`login` doesn't set the `isLoggedIn` flag: that's done by `handleincomingRedirect`. Basically, `Session` is a thin layer on top of `ClientAuthentication`, which is the core class here. Moreover, when calling `login` on a `ClientAuthentication` object, depending on the parameters, different behaviours are expected. Should all the possible behaviours be tested at this layer (e.g. the different OIDC redirect requests), knowing that these tests will be very redundant with the unit tests of the actual classes handling each of these flows (e.g. `AuthCodeWithPkceLoginHandler`?

Reply:
Ah ok. Well, if this is just a thin wrapper, and all the behaviour is actually in a wrapped class, then I'd just change this test to:
```
expect(await mySession.login({}).not.toThrow();
```
...given it returns (a promise) on `void`. If the method under test returned anything at all I'd probably do:
```
expect(await mySession.login({}).not.isNull();
```
...or something.
So yeah, I agree not to test the expected flows, since as you say, they'll all be tested elsewhere, but I don't think this test should be aware of the internal details of the wrapper class (i.e. that it's calling a specific method on the wrapped class) - the only behaviour of the wrapper to test is that it doesn't fail (since it doesn't have any **_behaviour_** of it's own to test really).

Reply:
Maybe it's being pedantic, but isn't a test just checking if something doesn't fail, as opposed to testing for actual output or for internal behaviour, a code smell ? As in that's just going to bump the code coverage, but it won't provide any value, and even worse in the case of a wrapper class like here it could hide the fact that the wrapped class isn't called at all (even if that's unlikely). What do you think ?

Reply:
> a code smell ?
No, I don't think so. Yes, it might be needed to get coverage (although maybe not in this case, as other tests will exercise the `login` method), but I do think it's still valid to test that a wrapper method doesn't fail when invoked with correct the params. I generally wouldn't **_also_** test for failure modes if I think the wrapped class itself has full coverage (and if the wrapper class really is **_just_** a wrapper (i.e. that it doesn't catch exceptions from the wrapped class and add it's own extra context to that exception for instance - but of course, if it **_did_** do that, then it **_would_** be adding behaviour, and therefore would absolutely have to be tested for here too)).